File: tests/benchmarks/resolution/fixtures/ruby/service.rb
require_relative "repository"
require_relative "validators"

def create_user(store, id, name, email)
  unless validate_user(name, email)
    raise ArgumentError, "Invalid user data"
  end
  user = { id: id, name: name, email: email }
  repo_save(store, id, user)
  user
end

def find_user(store, id)
  repo_find_by_id(store, id)
end

def find_user_by_email(store, email)
  repo_find_by_email(store, email)
end

def remove_user(store, id)
  repo_delete(store, id)
end

File: tests/benchmarks/resolution/fixtures/rust/service.rs
use crate::models::{create_user, User};
use crate::repository::{create_repository, UserRepository};
use crate::validator::validate_all;

pub struct UserService {
    repo: UserRepository,
}

impl UserService {
    pub fn new(repo: UserRepository) -> Self {
        UserService { repo }
    }

    pub fn get_user(&self, id: u64) -> Option<User> {
        self.repo.find_by_id(id)
    }

    pub fn add_user(&self, id: u64, name: &str, email: &str) -> Result<(), String> {
        let user = create_user(id, name, email);
        validate_all(&user)?;
        self.repo.save(&user)
    }

    pub fn remove_user(&self, id: u64) -> bool {
        self.repo.delete(id)
    }
}

pub fn build_service() -> UserService {
    let repo = create_repository();
    UserService::new(repo)
}
